### PR TITLE
Add fastMRI pytorch dataset

### PR DESCRIPTION
**Pull request by MaximeSongIdris**
Adding support for
- [x] single-coil MRI
- [x] multi-coil TODO

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by MaximeSongIdris** [comment]
In the actual iteration, since kspace of different mris have different (height, width) sizes, the dataset can't be feed to a dataloader. One solution is to filter mri slices by (height, width) size to keep only one size of mri slice.

File: deepinv/datasets/div2k.py [hunk 1]
```diff
@@ -30,65 +33,60 @@ class DIV2K(torch.utils.data.Dataset):
...
+                        f"""
+                    The train folder already exists,
+                    thus the download is aborted.
+                    Please set `download=False`
+                    OR remove `{train_dir_path}`."""
+                    )
+
+                if os.path.exists(valid_dir_path):
+                    raise ValueError(
+                        f"""
```
Comment (tachella) [inline deepinv/datasets/div2k.py:64]: I think there is a newline in the string that should be removed 

f"""text here 

instead of

f""" (new line)
text here

**Comment by tachella** [comment]
> In the actual iteration, since kspace of different mris have different (height, width) sizes, the dataset can't be feed to a dataloader. One solution is to filter mri slices by (height, width) size to keep only one size of mri slice.

@matthieutrs do you know more about this?
- Do they have different sizes because they correspond to different acceleration factors of the same underlying image size, or because the are related to different image sizes?
- can we incorporate a torchvision transform that resizes/normalizes?

File: deepinv/datasets/fastmri.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    :param callable, optional transform_kspace: A function/transform that takes in the
+        kspace and returns a transformed version. E.g, ``torchvision.transforms.RandomCrop``
+    :param callable, optional transform_target: A function/transform that takes in the
+        target and returns a transformed version. E.g, ``torchvision.transforms.RandomCrop``
+
+    |sep|
+
+    :Examples:
+
+        Instanciate dataset without transform ::
```
Comment (tachella) [inline deepinv/datasets/fastmri.py:84]: 1) Are we not testing on purpose here?

2) is this showing right on sphinx? I feel that the :: should be on a new line.

File: deepinv/datasets/fastmri.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        self.root --- file1000005.h5
+                   |
+                   -- xxxxxxxxxxx.h5
+
+    | 0) To download raw data, please go to the bottom of the page `https://fastmri.med.nyu.edu/`
+    | 1) Each MRI scan is stored in a HDF5 file and can be read with the h5py package.
+    | Each file contains the k-space data, ground truth and some meta data related to the scan.
+    | 2) MRI scans can either be single-coil or multi-coil with each coil in
+    | a multi-coil MRI scan focusses on a different region of the image.
+    | 3) In multi-coil MRIs, k-space has the following shape:
```
Comment (tachella) [inline deepinv/datasets/fastmri.py:52]: k-space --> the k-space data

File: deepinv/datasets/fastmri.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    | (number of slices, number of coils, height, width)
+    | 4) For single-coil MRIs, k-space has the following shape:
```
Comment (tachella) [inline deepinv/datasets/fastmri.py:54]: same here

File: deepinv/datasets/fastmri.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+    :param Union[str, Path] root: Path to the dataset.
+    :param bool test: Whether the split is the "test" set.
+    :param str challenge: "singlecoil" or "multicoil" depending on the type of mri scan.
+    :param bool load_metadata_from_cache: Whether to load dataset metadata from cache.
+    :param bool save_metadata_to_cache: Whether to cache dataset metadata.
+    :param Union[str, Path] metadata_cache_file: A file used to cache dataset
+        information for faster load times.
+    :param callable, optional sample_filter: A callable object that takes a
+        SliceSampleFileIdentifier as input and returns a boolean indicating
```
Comment (tachella) [inline deepinv/datasets/fastmri.py:65]: add a ref to the class with :meth:`deepinv.datasets.SliceSampleFileIdentifier` (check that this works)

File: deepinv/datasets/fastmri.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    | Each file contains the k-space data, ground truth and some meta data related to the scan.
+    | 2) MRI scans can either be single-coil or multi-coil with each coil in
+    | a multi-coil MRI scan focusses on a different region of the image.
+    | 3) In multi-coil MRIs, k-space has the following shape:
+    | (number of slices, number of coils, height, width)
+    | 4) For single-coil MRIs, k-space has the following shape:
+    | (number of slices, height, width)
+
+    :param Union[str, Path] root: Path to the dataset.
+    :param bool test: Whether the split is the "test" set.
```
Comment (tachella) [inline deepinv/datasets/fastmri.py:58]: This is a bit non-standard, I would add a train flag rather than test

File: deepinv/physics/blur.py [hunk 7]
```diff
@@ -380,6 +380,9 @@ def update_parameters(self, filter=None, **kwargs):
         if filter is not None:
             self.filter = torch.nn.Parameter(filter, requires_grad=False)
 
+        if hasattr(self.noise_model, "update_parameters"):
```
Comment (tachella) [inline deepinv/physics/blur.py:383]: why is this here? It should be part of another PR

File: deepinv/physics/mri.py [hunk 8]
```diff
@@ -22,86 +24,206 @@ class MRI(DecomposablePhysics):
...
+                  [ 0.4564, -0.0000]],
+        <BLANKLINE>
+                 [[ 0.0000,  1.8622],
+                  [ 0.0603, -0.0000]]]])
+        >>> physics = MRI(img_size=x.shape) # No subsampling
         >>> physics(x)
-        tensor([[-0.5305,  0.0351,  0.3326,  2.1730,  1.7072,  0.0418]])
+        tensor([[[[ 2.2908, -1.4290],
+                  [ 0.4564, -0.1814]],
+        <BLANKLINE>
```
Comment (tachella) [inline deepinv/physics/mri.py:54]: why the BLANKLINE?

File: deepinv/physics/mri.py [hunk 9]
```diff
@@ -22,86 +24,206 @@ class MRI(DecomposablePhysics):
...
-            )
-
-    def U_adjoint(self, x):
-        _, c, h, w = self.mask.shape
-        out = torch.zeros((x.shape[0], c, h, w), device=x.device)
+    def U(self, x: Tensor) -> Tensor:
+        if not torch.all((self.mask == 0) | (self.mask == 1)):
+            warnings.warn("The mask must have values 0 or 1")
 
-        if self.mask.size(0) == 1:
```
Comment (tachella) [inline deepinv/physics/mri.py:88]: why did you remove the size reduction? @matthieutrs

**Review by tachella — COMMENTED** [review]
I have a few minor comments. I worry that there are some changes that correspond to old PRs, is this expected?

File: deepinv/datasets/fastmri.py (hunk 6 above)
Comment (MaximeSongIdris) [reply deepinv/datasets/fastmri.py:58]: yes, here I want to separate the behavior between train/val vs test

File: deepinv/physics/mri.py (hunk 8 above)
Comment (MaximeSongIdris) [reply deepinv/physics/mri.py:54]: Duno, that's the result we get when testing with doctest :

python -m doctest -v deepinv/physics/mri.py

File: deepinv/datasets/fastmri.py (hunk 6 above)
Comment (tachella) [reply deepinv/datasets/fastmri.py:53]: why not a 'mode' which equals 'train' 'val' or 'test?

File: deepinv/datasets/fastmri.py (hunk 2 above)
Comment (tachella) [reply deepinv/datasets/fastmri.py:79]: perfect!

File: deepinv/datasets/fastmri.py (hunk 6 above)
Comment (MaximeSongIdris) [reply deepinv/datasets/fastmri.py:53]: train and val have return the target, while test doesn't
and we don't use mode for other

**Review by tachella — APPROVED** [review]
